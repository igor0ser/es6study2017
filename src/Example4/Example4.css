.example4 ul{
  list-style: none;
  padding: 0;
  display: flex;
  flex-wrap: wrap;
  max-width: 475px;
  cursor: pointer;
  justify-content: center;
}

.example4 li{
  border: 1px dashed gold;
  text-align: center;
  margin: 3px;
  padding: 3px;
}