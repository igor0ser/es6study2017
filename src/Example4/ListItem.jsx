import React, { Component } from 'react';

class ListItem extends Component {
/*  shouldComponentUpdate(nextProps) {
    return this.props === nextProps;
  }*/

  render() {
    console.log('render list item');
    const { item } = this.props;

    return (
      <li>
        { item }
      </li>
    );
  }
}

export default ListItem;
