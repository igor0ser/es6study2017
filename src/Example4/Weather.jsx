import React from 'react';

const Weather = ({
  name, weather, main: { temp_min, temp_max }
}) => (
    <div style={{ margin: 20 }}>
        <strong>{name}</strong>
        : {weather[0].description}, {temp_min} - {temp_max} Â°C
    </div>
);

export default Weather;
