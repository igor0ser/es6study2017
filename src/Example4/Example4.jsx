import React, { Component } from 'react';
import lodash from 'lodash';
import ListItem from './ListItem';

import './Example4.css';

class Example4 extends Component {
    constructor(props) {
        super(props);

        this.state = {
            data: lodash.range(1, 101),
            ascending: true
        };
    }

    sortArray = () => {
      const { data, ascending } = this.state;

      const newAscending = !ascending;
      const coef = newAscending ? 1 : -1;
      const sortedData = [...data].sort((a, b) => coef * (a - b));

      this.setState({
        ascending: newAscending,
        data: sortedData
      })
    }

    render() {
        const { data } = this.state;

        return (
            <section className="example4">
              <h3>Example 4</h3>
              <ul onClick={this.sortArray}>
                {data.map(item => <ListItem item={item} key={item} />)}
              </ul>
            </section>
        );
    }
}

export default Example4;
