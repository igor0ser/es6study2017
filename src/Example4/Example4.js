import React, { Component } from 'react';
import axios from 'axios';
import Weather from './Weather';
import buildUrl from './buildUrl';


class Example4 extends Component {
  constructor(props) {
    super(props);

    this.state = {
      query: 'Kiev',
      data: null
    };
  }

  componentDidMount() {
    this.fetchData();
  }

  fetchData = () => {
    axios.get(buildUrl(this.state.query))
      .then(({ data }) => this.setState({ data }));
  }

  searchCity = (e) => {
    if(e.key !== 'Enter') return;

    this.setState({
      query: e.target.value,
      data: null
    }, this.fetchData)
  }

  render() {
    const { data, query } = this.state;
    console.log(data);

    return (
      <section>
        <h3>Example 4</h3>
        <input
          type="text"
          defaultValue={query}
          onKeyPress={this.searchCity}
        />
        {data ? <Weather {...data} /> : 'Loading...'}
      </section>
    );
  }
}

export default Example4;
