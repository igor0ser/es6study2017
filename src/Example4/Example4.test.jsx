import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import axios from 'axios';

import Example4 from './Example4';

jest.mock('./buildUrl', () =>
    cityName => cityName
);

const fakeData1 = {
    name: 'cityName1',
    weather: [{ description: 'clear sky'}],
    main: { temp_min: 10, temp_max: 20 },
};

const fakeData2 = {
    name: 'cityName2',
    weather: [{ description: 'clear sky'}],
    main: { temp_min: 10, temp_max: 20 },
};

describe('Example1', () => {
    let fakeData
    let comp;
    let response1;
    let response2;

    beforeEach(() => {
        response1 = new Promise(resolve => resolve({ data: fakeData1 }));
        response2 = new Promise(resolve => resolve({ data: fakeData2 }));

        axios.get = jest.fn()
                .mockReturnValueOnce(response1)
                .mockReturnValueOnce(response2);
    });

    it('should render loading before data is loaded and fetch data', () => {
        const comp = mount(<Example4 />);
        expect(comp.find('.content').text()).toEqual('Loading...');
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('should render Weather widget after data is loaded', () => {
        const comp = mount(<Example4 />);
        return response1.then(() => {
            expect(axios.get.mock.calls.length).toBe(1);
            expect(axios.get.mock.calls[0]).toEqual(['Kiev']);

            expect(comp.find('Weather').exists()).toEqual(true);
            expect(toJson(comp)).toMatchSnapshot();
        })
    });

    it('should rerender Weather widget after new data is loaded', () => {
        comp = mount(<Example4 />);

        const input = comp.find('input');
        input.get(0).value = 'dummy';

        input.simulate('keyPress', { key: 'Enter' })
        expect(comp.state()).toEqual({
            query: 'dummy',
            data: null
        });

        return response2.then(() => {
            expect(axios.get.mock.calls.length).toBe(2);
            expect(axios.get.mock.calls[1]).toEqual(['dummy']);
            expect(comp.find('Weather').props().name).toEqual('cityName2');
            expect(toJson(comp)).toMatchSnapshot();
        });
    });
});
