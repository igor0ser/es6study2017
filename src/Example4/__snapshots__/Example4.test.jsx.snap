// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example1 should render Weather widget after data is loaded 1`] = `
<Example4>
  <section>
    <h3>
      Example 4
    </h3>
    <input
      defaultValue="Kiev"
      onKeyPress={[Function]}
      type="text"
    />
    <b
      className="content"
    >
      <Weather
        main={
          Object {
            "temp_max": 20,
            "temp_min": 10,
          }
        }
        name="cityName1"
        weather={
          Array [
            Object {
              "description": "clear sky",
            },
          ]
        }
      >
        <div
          style={
            Object {
              "margin": 20,
            }
          }
        >
          <strong>
            cityName1
          </strong>
          : 
          clear sky
          , 
          10
           - 
          20
           °C
        </div>
      </Weather>
    </b>
  </section>
</Example4>
`;

exports[`Example1 should render loading before data is loaded and fetch data 1`] = `
<Example4>
  <section>
    <h3>
      Example 4
    </h3>
    <input
      defaultValue="Kiev"
      onKeyPress={[Function]}
      type="text"
    />
    <b
      className="content"
    >
      Loading...
    </b>
  </section>
</Example4>
`;

exports[`Example1 should rerender Weather widget after new data is loaded 1`] = `
<Example4>
  <section>
    <h3>
      Example 4
    </h3>
    <input
      defaultValue="dummy"
      onKeyPress={[Function]}
      type="text"
    />
    <b
      className="content"
    >
      <Weather
        main={
          Object {
            "temp_max": 20,
            "temp_min": 10,
          }
        }
        name="cityName2"
        weather={
          Array [
            Object {
              "description": "clear sky",
            },
          ]
        }
      >
        <div
          style={
            Object {
              "margin": 20,
            }
          }
        >
          <strong>
            cityName2
          </strong>
          : 
          clear sky
          , 
          10
           - 
          20
           °C
        </div>
      </Weather>
    </b>
  </section>
</Example4>
`;
